File: solutions/blind-75/string/validParentheses.js
const isValid = (string) => {
    const stack = [];
    const correspondingParentheses = {'(': ')', '{': '}', '[': ']'};
    
    for (let i = 0; i < string.length; i++) {
        const char = string[i];

        const isOpeningParentheses = correspondingParentheses.hasOwnProperty(char);
        if (isOpeningParentheses) {
            stack.push(char);
            continue;
        }
        
        const lastChar = stack.pop();
        const isValidClosingParantheses = correspondingParentheses[lastChar] === char;
        if (!isValidClosingParantheses) return false;
    }
    
    const isStackEmpty = stack.length === 0;
    return isStackEmpty
};

File: solutions/python/blind-75/string/valid_parentheses.py
class Solution:
    def isValid(self, string: str) -> bool:
        stack = []
        corresponding_parentheses = {
            '(': ')',
            '{': '}',
            '[': ']',
        }
        
        for char in string:
            is_opening_parentheses = char in corresponding_parentheses

            if is_opening_parentheses:
                stack.append(char)
                continue
                
            if len(stack) == 0:
                return False
            
            last_char = stack.pop()
            
            is_valid_closing_parentheses = corresponding_parentheses[last_char] == char
            if not is_valid_closing_parentheses:
                return False
        
        is_stack_empty = len(stack) == 0
        return is_stack_empty